src/libraries/Microsoft.Extensions.DependencyInjection/src/ServiceLookup/ServiceProviderEngineScope.cs

```diff
@@ -187,12 +187,17 @@ static async ValueTask Await(int i, ValueTask vt, List<object> toDispose)
                 // No further changes to _state.Disposables, are allowed.
                 _disposed = true;
 
-                // ResolvedServices is never cleared for singletons because there might be a compilation running in background
-                // trying to get a cached singleton service. If it doesn't find it
-                // it will try to create a new one which will result in an ObjectDisposedException.
+            }
 
-                return _disposables;
+            if (IsRootScope)
+            {
+                RootProvider.Dispose();
```

Comment:
Are we OK with the extra loop a Dispose will cause now?

Before:
`ServiceProvider.Dispose()` => `ServiceProviderEngineScope.Dispose()`

Now:
`ServiceProvider.Dispose()` => `ServiceProviderEngineScope.Dispose()` => `ServiceProvider.Dispose()` => `ServiceProviderEngineScope.Dispose()` (which finally breaks out of the cycle.

We could fix this (if necessary) by checking `if (IsRootScope && !RootProvider._disposed)`.

Reply:
I thought about that and didn't like to change _disposed to be not private. I don't have a strong feeling though. If you think is worth it I can apply it.

Reply:
Thinking more, I think it is ok to introduce IsDisposed. I'll do that.